### ISSUE TITLE
[java] UselessOverridingMethod - possible false negative with package-private class finalize

### ISSUE THREAD

**Issue by adangel** [pmd/pmd#2879]
**Affects PMD Version:** 6.30.0-SNAPSHOT

**Rule:** [UselessOverridingMethod](https://pmd.github.io/latest/pmd_rules_java_design.html#uselessoverridingmethod)

**Description:**

From https://github.com/pmd/pmd/pull/2871#issuecomment-717848659

Note: There seems to be a difference when having the class in the auxclasspath or not (typeresolution).

**Code Sample demonstrating the issue:**

```java
class SuperFinalizer {
    protected void finalize() throws Throwable { //should warn: Overriding method merely calls super 
        super.finalize();
    }
}

```

**Expected outcome:**

*   Is PMD missing to report a violation, where there should be one? -> false-negative

**Running PMD through:** regression tester

**Comment by HoussemNasri** [comment]
Hi @adangel, I would like to make my first contribution to this project, I came across this issue and it seems interesting, please let me know if it's still open.

**Comment by adangel** [comment]
Hi @HoussemNasri , thanks and welcome!
Yes, the issue is still open.

**Comment by cchang83** [comment]
Hello @adangel, I tried to reproduce this bug by copying your above code (class SuperFinalizer.java) and running it, but I seem to produce the expected outcome instead of the bug. I am getting the output "overriding method merely calls super."

Do you have other tests or scenarios that reproduce this bug?

**Comment by adangel** [comment]
I @cchang83 , you can reproduce the bug with the above code snippet, if you
1. compile the file
2. use  the [auxclasspath](https://pmd.github.io/latest/pmd_userdocs_cli_reference.html#auxclasspath) option

E.g.

```
mkdir testcase-2879 && cd testcase-2879
mkdir src bin
echo 'class SuperFinalizer {
    protected void finalize() throws Throwable { //should warn: Overriding method merely calls super 
        super.finalize();
    }
}' > src/SuperFinalizer.java
javac -d bin src/SuperFinalizer.java
pmd-bin-6.40.0/bin/run.sh pmd -d src -auxclasspath bin -f text -no-cache -R category/java/design.xml/UselessOverridingMethod
# no output

# without auxclasspath
pmd-bin-6.40.0/bin/run.sh pmd -d src -f text -no-cache -R category/java/design.xml/UselessOverridingMethod
/home/andreas/temp/testcase-2879/src/SuperFinalizer.java:2:	UselessOverridingMethod:	Overriding method merely calls super
```

The goal is, that this rule also produces the violation with the auxclasspath correctly filled.

**Comment by jsotuyod** [comment]
This is down to this block of code:

https://github.com/pmd/pmd/blob/c9926fba4c24627237852aa968e6368c217b9b79/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/UselessOverridingMethodRule.java#L109-L119

We detect the override is only calling super, but when we check if the visibility is being changed (valid use of such overrides), we see that modifiers are the same (`protected`) BUT, since the class is in a different package, this means we are actually opening up the visibility (other classes may now invoke it on this class). This is perfectly correct.

It seems that we either make a special case for `finalize`, or we keep ignoring it now that the whole finalization process is deprecated for removal, and just wait for the problem to go away on it's own.

**Comment by adangel** [comment]
> since the class is in a different package, this means we are actually opening up the visibility (other classes may now invoke it on this class).

I'm pretty sure, I did _not_ think about that, when I created the issue. That this is about `finalize` is just a coincidence.

So, this is really just a valid case then (in terms of: the rule is correct in not reporting this), isn't it?

**Comment by jsotuyod** [comment]
> So, this is really just a valid case then (in terms of: the rule is correct in not reporting this), isn't it?

Yes, it is. Would you rather close this?
